ra/ra.go

```diff
@@ -242,22 +244,65 @@ type certificateRequestEvent struct {
 // registration-based overrides are necessary.
 const noRegistrationID = -1
 
-func (ra *RegistrationAuthorityImpl) checkRegistrationLimit(ctx context.Context, ip net.IP) error {
-	limit := ra.rlPolicies.RegistrationsPerIP()
+// registrationCounter is a type to abstract the use of
+// ra.SA.CountRegistrationsByIP or ra.SA.CountRegistrationsByIPRange
+type registrationCounter func(context.Context, net.IP, time.Time, time.Time) (int, error)
 
-	if limit.Enabled() {
-		now := ra.clk.Now()
-		count, err := ra.SA.CountRegistrationsByIP(ctx, ip, limit.WindowBegin(now), now)
-		if err != nil {
-			return err
-		}
-		if count >= limit.GetThreshold(ip.String(), noRegistrationID) {
-			ra.regByIPStats.Inc("Exceeded", 1)
-			ra.log.Info(fmt.Sprintf("Rate limit exceeded, RegistrationsByIP, IP: %s", ip))
-			return berrors.RateLimitError("too many registrations for this IP")
-		}
-		ra.regByIPStats.Inc("Pass", 1)
+// checkRegistrationIPLimit checks a specific registraton limit by using the
+// provided registrationCounter function to determine if the limit has been
+// exceeded for a given IP or IP range
+func (ra *RegistrationAuthorityImpl) checkRegistrationIPLimit(
+	ctx context.Context,
+	limit ratelimit.RateLimitPolicy,
+	ip net.IP,
+	counter registrationCounter) error {
+
+	if !limit.Enabled() {
+		return nil
 	}
+
+	now := ra.clk.Now()
+	windowBegin := limit.WindowBegin(now)
+	count, err := counter(ctx, ip, windowBegin, now)
+	if err != nil {
+		return err
+	}
+
+	if count >= limit.GetThreshold(ip.String(), noRegistrationID) {
+		return berrors.RateLimitError("too many registrations for this IP")
+	}
+
+	return nil
+}
+
+// checkRegistrationLimits enforces the RegistrationsPerIP and
+// RegistrationsPerIPRange limits
+func (ra *RegistrationAuthorityImpl) checkRegistrationLimits(ctx context.Context, ip net.IP) error {
+	// Check the registrations per IP limit using the CountRegistrationsByIP SA
+	// function that matches IP addresses exactly
+	exactRegLimit := ra.rlPolicies.RegistrationsPerIP()
+	err := ra.checkRegistrationIPLimit(ctx, exactRegLimit, ip, ra.SA.CountRegistrationsByIP)
+	if err != nil {
+		ra.regByIPStats.Inc("Exceeded", 1)
+		ra.log.Info(fmt.Sprintf("Rate limit exceeded, RegistrationsByIP, IP: %s", ip))
+		return err
+	}
+	ra.regByIPStats.Inc("Pass", 1)
+
+	// Check the registrations per IP range limit using the
```

Comment:
I thought potentially we might want to add an IPv4 range limit in the future, but that's an extremely weak justification for increasing DB load. I agree with you - I think this can be updated to ret early. Thanks!